op-batcher/batcher/tx_data.go

```diff
@@ -12,47 +14,100 @@ import (
+	}
+	return id
 }
 
 // Bytes returns the transaction data. It's a version byte (0) followed by the
 // concatenated frames for this transaction.
+// TODO(Seb) rename to CallData
 func (td *txData) Bytes() []byte {
-	return append([]byte{derive.DerivationVersion0}, td.frame.data...)
+	data := make([]byte, 1, 1+td.Len())
+	data[0] = derive.DerivationVersion0
+	for _, f := range td.frames {
+		data = append(data, f.data...)
+	}
+	return data
+}
+
+func (td *txData) Blobs() ([]*eth.Blob, error) {
+	blobs := make([]*eth.Blob, 0, len(td.frames))
+	for _, f := range td.frames {
+		var blob eth.Blob
+		if err := blob.FromData(append([]byte{derive.DerivationVersion0}, f.data...)); err != nil {
+			return nil, err
+		}
+		blobs = append(blobs, &blob)
+	}
+	return blobs, nil
 }
 
-func (td *txData) Len() int {
-	return 1 + len(td.frame.data)
+// TODO(Seb) check all uses of Len for right dimension
+func (td *txData) Len() (l int) {
+	for _, f := range td.frames {
+		l += len(f.data)
+	}
+	return l
 }
 
 // Frame returns the single frame of this tx data.
-//
-// Note: when the batcher is changed to possibly send multiple frames per tx,
-// this should be changed to a func Frames() []frameData.
-func (td *txData) Frame() frameData {
-	return td.frame
+func (td *txData) Frames() []frameData {
+	return td.frames
 }
 
 // txID is an opaque identifier for a transaction.
-// It's internal fields should not be inspected after creation & are subject to change.
-// This ID must be trivially comparable & work as a map key.
-//
-// Note: transactions currently only hold a single frame, so it can be
-// identified by the frame. This needs to be changed once the batcher is changed
-// to send multiple frames per tx.
-type txID = frameID
+// Its internal fields should not be inspected after creation & are subject to change.
+// Its String() can be used for comparisons and works as a map key.
+type txID []frameID
 
 func (id txID) String() string {
-	return fmt.Sprintf("%s:%d", id.chID.String(), id.frameNumber)
+	var (
+		sb      strings.Builder
+		curChID derive.ChannelID
+	)
+	for _, f := range id {
+		if f.chID == curChID {
+			sb.WriteString(fmt.Sprintf("+%d", f.frameNumber))
+		} else {
+			if curChID != (derive.ChannelID{}) {
+				sb.WriteString("|")
+			}
+			curChID = f.chID
+			sb.WriteString(fmt.Sprintf("%s:%d", f.chID.String(), f.frameNumber))
+		}
+	}
+	return sb.String()
 }
 
 // TerminalString implements log.TerminalStringer, formatting a string for console
 // output during logging.
 func (id txID) TerminalString() string {
-	return fmt.Sprintf("%s:%d", id.chID.TerminalString(), id.frameNumber)
+	var (
+		sb      strings.Builder
+		curChID derive.ChannelID
+	)
+	for _, f := range id {
+		if f.chID == curChID {
+			sb.WriteString(fmt.Sprintf("+%d", f.frameNumber))
+		} else {
+			if curChID != (derive.ChannelID{}) {
+				sb.WriteString("|")
+			}
+			curChID = f.chID
+			sb.WriteString(fmt.Sprintf("%s:%d", f.chID.TerminalString(), f.frameNumber))
```

Comment:
I guess this line is the only difference between TerminalString() and String()?  Perhaps remove the duplicate code by having a generic method that takes a functor to compute the chID string?

Reply:
Yeah I thought about that, but was too lazy at the time. Will optimize this once everything else works.

Reply:
done